core/lib/zksync_core/src/api_server/web3/state.rs

```diff
@@ -543,10 +546,53 @@ impl<E> RpcState<E> {
 /// Contains mapping from index to `Filter` with optional location.
 #[derive(Default, Debug, Clone)]
 pub struct Filters {
-    state: HashMap<U256, TypedFilter>,
+    state: HashMap<U256, InstalledFilter>,
     max_cap: usize,
 }
 
+#[derive(Debug, Clone)]
+struct InstalledFilter {
+    pub filter: TypedFilter,
+    #[allow(dead_code)]
+    guard: Arc<GaugeGuard>,
+    created_at: Instant,
+    last_request: u64,
+    request_count: usize,
+}
+
+impl InstalledFilter {
+    pub fn new(filter: TypedFilter) -> Self {
+        let guard = FILTER_METRICS.metrics_count[&FilterType::from(&filter)].inc_guard(1);
+        Self {
+            filter,
+            guard: Arc::new(guard),
+            created_at: Instant::now(),
+            last_request: chrono::Utc::now().naive_utc().timestamp() as u64,
+            request_count: 0,
+        }
+    }
+
+    pub fn update_stats(&mut self) {
+        let previous_request_timestamp = self.last_request;
+        let now = chrono::Utc::now().naive_utc().timestamp() as u64;
+
+        self.last_request = now;
+        self.request_count.add_assign(1);
+
+        let filter_type = FilterType::from(&self.filter);
+        FILTER_METRICS.request_frequency[&filter_type]
+            .observe(Duration::from_secs(now - previous_request_timestamp));
+    }
+}
+
+impl Drop for InstalledFilter {
+    fn drop(&mut self) {
+        FILTER_METRICS.filter_count[&FilterType::from(&self.filter)].observe(self.request_count);
```

Comment:
Nit: you can extract type into a local variable to not create it twice (the compiler may optimize it away, but it's easier to read in any case).